### ISSUE TITLE
Image Not Detected

### ISSUE THREAD

**Issue by cinnamolly** [serratus/quaggaJS#119]
Hi,

I'm finding that static images with clear barcodes (even barcodes generated from the Internet) are either resulting in "not detected" or the error:
`if(result.codeResult) {
             ^
TypeError: Cannot read property 'codeResult' of undefined`

Is there a certain type or resolution of image that needs to be uploaded? Why does this happen to my static images, but not the test images?

Any help would be greatly appreciated.

Thank you,
Molly

**Comment by cinnamolly** [comment]
For example, this image 
![bar2](https://cloud.githubusercontent.com/assets/9923980/16880815/1ec28a32-4a85-11e6-9127-e0453c35cca3.jpg)
returned the error I pasted above. Attached additionally is my source code.
[bar.txt](https://github.com/serratus/quaggaJS/files/366405/bar.txt)

**Comment by sidonaldson** [comment]
I have found this too. It would appear that if it doesn't find a barcode it returns the callback without  a `result` argument.

On line:4398 of lib/quagga.js I added it to return a empty object if result was null
```
           this.init(config, function () {
	            _events2.default.once("processed", function (result) {
	                _this.stop();
	                resultCallback.call(null, result || {});
	            }, true);
	            _start();
	        });
```

**Comment by TomasHubelbauer** [comment]
I agree with this not being a good experience, I also think it happened to me before. Definitely something we should try to track down and prevent.

**Comment by ericblade** [comment]
While getting testing working standalone with #303, I noticed that the callback is called pretty frequently without the expected code data.  The question I guess I have, is are my expectations incorrect, or is that an incorrect response?  decodeSingle() probably should callback with *something*.  However, there seems to be an expectation throughout the code base that result codes will always be available to the callback ..

**Comment by TomasHubelbauer** [comment]
I'll need to take a closer look at the internal structure to understand how we could improve this, but as far as expectations go, I think the current design could be improved.

**Comment by ericblade** [comment]
What I see in src/quagga.js in the area of line 534 is

````
        this.init(config, () => {
            Events.once("processed", (result) => {
                this.stop();
                resultCallback.call(null, result);
            }, true);
            start();
        });
````

.... and here-in lies the specific problem:

decodeSingle's callback is *not* called similarly to onDetected -- it is called similarly to onProcessed.  

In src/quagga.js publishResult()

````
function publishResult(result, imageData) {
    let resultToPublish = result;

    if (result && _onUIThread) {
        transformResult(result);
        addResult(result, imageData);
        resultToPublish = result.barcodes || result;
    }

    Events.publish("processed", resultToPublish);
    if (hasCodeResult(result)) {
        Events.publish("detected", resultToPublish);
    }
}
````

So, decodeSingle calls resultCallback as if it were onProcessed, but existing code (including at least the tests, and probably the samples, and the README, but i haven't looked yet) is expecting it to be called as if it were onDetected.

I think this could cause quite a few questions.

**Comment by ericblade** [comment]
Also, in my first comment on this thread, i had said onDetected, without reading the code file.  The people reporting this problem have been using decodeSingle().  I'm pretty sure that in live mode, I've never seen onDetected called without a result, even if it isn't an accurate result.  

SO @cinnamolly and @sidonaldson if you're still out here, the root of the problem is expecting decodeSingle() to always return a codeResult -- if it wasn't able to decode a barcode, then you get no codeResult.
